test/hotspot/jtreg/compiler/lib/template_framework/Renderer.java

```diff
@@ -0,0 +1,398 @@
+    }
+
+    List<Name> listNames(NameSet.Predicate predicate) {
+        return currentCodeFrame.listNames(predicate);
+    }
+
+    /**
+     * Formats values to {@link String} with the goal of using them in Java code.
+     * By default we use the overrides of {@link Object#toString}.
+     * But for some boxed primitives we need to create a special formatting.
+     */
+    static String format(Object value) {
+        return switch (value) {
+            case String s -> s;
+            case Integer i -> i.toString();
+            // We need to append the "L" so that the values are not interpreted as ints,
+            // and then javac might complain that the values are too large for an int.
+            case Long l -> l.toString() + "L";
+            // Some Float and Double values like Infinity and NaN need a special representation.
+            case Float f -> formatFloat(f);
+            case Double d -> formatDouble(d);
+            default -> value.toString();
+        };
+    }
+
+    private static String formatFloat(Float f) {
+        if (Float.isFinite(f)) {
+            return f.toString() + "f";
+        } else if (f.isNaN()) {
+            return "Float.intBitsToFloat(" + Float.floatToRawIntBits(f) + " /* NaN */)";
+        } else if (f.isInfinite()) {
+            if (f > 0) {
+                return "Float.POSITIVE_INFINITY";
+            } else {
+                return "Float.NEGATIVE_INFINITY";
+            }
+        } else {
+            throw new RuntimeException("Not handled: " + f);
+        }
+    }
+
+    private static String formatDouble(Double d) {
+        if (Double.isFinite(d)) {
+            return d.toString();
+        } else if (d.isNaN()) {
+            return "Double.longBitsToDouble(" + Double.doubleToRawLongBits(d) + "L /* NaN */)";
+        } else if (d.isInfinite()) {
+            if (d > 0) {
+                return "Double.POSITIVE_INFINITY";
+            } else {
+                return "Double.NEGATIVE_INFINITY";
+            }
+        } else {
+            throw new RuntimeException("Not handled: " + d);
+        }
+    }
+
+    private void renderTemplateToken(TemplateToken templateToken) {
+        TemplateFrame templateFrame = TemplateFrame.make(currentTemplateFrame, nextTemplateFrameId++);
+        currentTemplateFrame = templateFrame;
+
+        templateToken.visitArguments((name, value) -> addHashtagReplacement(name, format(value)));
+        TemplateBody body = templateToken.instantiate();
+        renderTokenList(body.tokens());
+
+        if (currentTemplateFrame != templateFrame) {
+            throw new RuntimeException("Internal error: TemplateFrame mismatch!");
+        }
+        currentTemplateFrame = currentTemplateFrame.parent;
+    }
+
+    private void renderToken(Token token) {
+        switch (token) {
+            case StringToken(String s) -> {
+                renderStringWithDollarAndHashtagReplacements(s);
+            }
```

Comment:
```suggestion
            case StringToken(String s) -> renderStringWithDollarAndHashtagReplacements(s);
```

Reply:
I think I prefer the uniformity of the brackets as I have it. Would that be ok for you too?